ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java

```diff
@@ -453,6 +463,56 @@ private void cleanupOldState() {
     localCommands.ifPresent(lc -> lc.processLocalCommandFiles(serviceContext));
   }
 
+  public void cleanupOldStateDirectories(final KsqlConfig configWithApplicationServer) {
+    final String stateDir = configWithApplicationServer.getKsqlStreamConfigProps().getOrDefault(
+            StreamsConfig.STATE_DIR_CONFIG,
+            StreamsConfig.configDef().defaultValues().get(StreamsConfig.STATE_DIR_CONFIG)).toString();
+
+    final Set<String> stateStoreNames =
+        ksqlEngine.getPersistentQueries()
+            .stream()
+            .map(PersistentQueryMetadata::getQueryApplicationId)
+            .collect(Collectors.toSet());
+    try {
+      Files.walkFileTree(Paths.get(stateDir), new SimpleFileVisitor<Path>() {
+        @Override
+        public FileVisitResult visitFileFailed(Path path, IOException exc) {
+          log.error("Error cleaning up obsolete state directories \n", exc);
+          return FileVisitResult.CONTINUE;
+        }
+
+        @Override
+        public FileVisitResult visitFile(Path path, BasicFileAttributes attrs) {
+          if (!stateStoreNames.contains(path.getFileName())) {
+            try {
+              Files.delete(path);
+              log.warn("Deleted local state store for non-existing query {}. This is not expected and was likely due to a " +
+                      "race condition when the query was dropped before.", path.getFileName());
```

Comment:
I think we need to be forward-looking a bit since as we binpack queries into KS, the local state dir names may not be the query ids but named topology ids (cc @ableegoldman @wcarlson5 could you clarify)?